frontend/ui/widget/textboxwidget.lua

```diff
@@ -2186,4 +2176,247 @@ function TextBoxWidget:getSourceIndex(char_idx)
     end
 end
 
+function TextBoxWidget:getPosFromAbsPos(abs_pos)
+    local pos = Geom:new{
+        x = abs_pos.x - self.dimen.x,
+        y = abs_pos.y - self.dimen.y,
+    }
+
+    -- check if the coordinates are actually inside our area
+    if pos.x < 0 or pos.x >= self.dimen.w or pos.y < 0 or pos.y >= self.dimen.h then
+        return nil
+    end
+
+    return pos
+end
+
+function TextBoxWidget:getXtextHighlightIndices(start_x, start_y, end_x, end_y)
+    -- With xtext and fribidi, words may not be laid out in logical order,
+    -- so the left of a visual word may be its end in logical order,
+    -- and the right its start.
+    -- So, just find out charpos (text indice) of both points and
+    -- find word edges in the logical order text/charlist.
+    local sel_start_idx = self:getCharPosAtXY(start_x, start_y)
+    local sel_end_idx = self:getCharPosAtXY(end_x, end_y)
+    if not sel_start_idx or not sel_end_idx then
+        -- one or both hold points were out of text
+        return nil, nil, nil, nil
+    end
+    if sel_start_idx > sel_end_idx then -- re-order if needed
+        sel_start_idx, sel_end_idx = sel_end_idx, sel_start_idx
+    end
+    -- We get cursor positions, which can be after last char,
+    -- and that we need to correct. But if both positions are
+    -- after last char, the full selection is out of text.
+    if sel_start_idx > #self._xtext then -- Both are after last char
+        return nil, nil, nil, nil
+    end
+    if sel_end_idx > #self._xtext then -- Only end is after last char
+        sel_end_idx = #self._xtext
+    end
+
+    local start_line_num = math.ceil(start_y / self.line_height_px) + self.virtual_line_num - 1
+    local end_line_num = math.ceil(end_y / self.line_height_px) + self.virtual_line_num - 1
+    if start_line_num < 1 or end_line_num > #self.vertical_string_list then
+        return nil, nil, nil, nil
```

Comment:
(I don't remember how all this works when we start or end text selection in a margin - and if you changed anything to this behaviour. I rarely do text selection in these text/htmlbox widgets, but in Reader, I often end them in the right or botto margin, so some people may be used to do it like me.)

Reply:
Yes, it works better in Reader. I will try to match its behavior.